### PR TITLE
Tuplet replace note

### LINKED ISSUE #72: Implement Tuplet.replaceNote methods
Many of the `Tuplet.replaceNote` methods are unimplemented. This issue is to implement those methods and their tests. Some of the test methods have been created an not implemented. Others have been just outlined. Each replaceNote method will call into the related `NoteCollection` version which is not `public`. The tests should cover only the public methods and not the non-public ones as well.

### PR DESCRIPTION
**Pull request by drumnkyle**
This fixes #72. 

- [x] Fix removeNotes method
- [x] Implement tests for `Tuplet.replaceNotes(in:with:[T])`
- [x] Handle ties

### CODE REVIEW

**Comment by drumnkyle** [comment]
I hit a tough bug to fix that seems like it will be complicated. If you could take a look at the TODO in the code and the OneNote document and give me any help, that would be great. Otherwise, I will come back to it another day.

**Comment by drumnkyle** [comment]
This is ready to go. If @robtimp, @migue48 could take a look that would be great. I will merge it at some point probably either way, because it's a lot of code and I understand if you don't review it ;)

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 1]
```diff
@@ -200,6 +481,11 @@ public struct Tuplet: NoteCollection {
 
 extension Tuplet: Equatable {
     public static func ==(lhs: Tuplet, rhs: Tuplet) -> Bool {
+        guard lhs.noteTimingCount == rhs.noteTimingCount &&
+            lhs.noteDuration == rhs.noteDuration &&
+            lhs.noteCount == rhs.noteCount else {
+                return false
+        }
         guard lhs.notes.count == rhs.notes.count else {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Do we want to compare each note as well?

File: MusicNotationKit/MusicNotationKit/Measure.swift [hunk 2]
```diff
@@ -65,7 +65,7 @@ public struct Measure: ImmutableMeasure, Equatable {
             assertionFailure("note collection should be tuplet, but cast failed")
             throw MeasureError.internalError
         }
-        try tuplet.replaceNote(at: tupletIndex, with: note)
+        try tuplet.replaceNote(at: tuplet.flatIndexes[tupletIndex], with: note)
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Measure.swift]: I am going to guess this is done this way due to the dependency in modifyTie? Leaving this comment so that we can come back here and update the code once we implement the internal version of `replaceNote(at:)`  at the Measure level.

File: MusicNotationKit/MusicNotationKit/NoteCollection.swift [hunk 3]
```diff
@@ -25,6 +25,9 @@ public protocol NoteCollection {
      */
     var noteTimingCount: Int { get }
 
+    var first: Note { get }
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/NoteCollection.swift]: We should probably add documentation to `first` and `last` mentioning that the library should die if `NoteCollection` is empty.  Is throwing `NoteCollectionError.Empty` an overkill right now?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 4]
```diff
@@ -18,13 +18,27 @@ public struct Tuplet: NoteCollection {
...
+    public let noteCount: Int
     /// The duration of the notes that define this tuplet
     public let noteDuration: NoteDuration
     /// The number of notes that this tuplet fits in the space of
     public let noteTimingCount: Int
+    public var first: Note {
+        do {
+            return try note(at: 0)
+        } catch {
+            fatalError(String(describing: error))
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: This seems like a one off case. We usually let the user handle the errors.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 5]
```diff
@@ -60,6 +74,7 @@ public struct Tuplet: NoteCollection {
         guard count > 1 else {
             throw TupletError.countMustBeLargerThan1
         }
+        noteCount = count
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: If `noteCount = count`, then it is probably better to keep `noteCount:Int { return notes.count}` or something like that?  I am just concerned about cases in which we replace a note with a couple of notes, and the note count needs to be updated as a result?  I am not sure if I am following correctly.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 6]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
         return try note(from: fullIndexes)
     }
 
-    public mutating func replaceNote(at index: Int, with note: Note) throws {
-        try replaceNote(at: index, with: note as NoteCollection)
+    public mutating func replaceNote<T: NoteCollection>(at index: Int, with noteCollection: T) throws {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: It seems that `replaceNote` in the general case can be used to replace a `Note` or a `Tuplet` or a range of noteCollection instances. I am not sure if  renaming to `replace(at index, with noteCollection: )` would be descriptive enough?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 7]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+        }
+        let flatIndex = flatIndexes[index]
+        let preservedTieStateCollection = try preserveTieStateForReplacement(in: index...index, with: noteCollection)
+        try replaceNote(at: flatIndex, with: preservedTieStateCollection)
     }
 
-    public mutating func replaceNote(at index: Int, with notes: [Note]) throws {
-        throw NSError()
+    public mutating func replaceNote<T: NoteCollection>(at index: Int, with noteCollections: [T]) throws {
+        // validate they are the same duration
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: probably don't need this comment (same line 136). If it is confusing, then maybe change the method's name?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 8]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+        // Need to translate the index now that notes have been removed
+        if toModify.flatIndexes.count > firstNoteIndex {
+            try toModify.insert(noteCollections, at: toModify.flatIndexes[firstNoteIndex])
+        } else {
+            // Just append
+            toModify.notes.append(contentsOf: noteCollections)
+        }
+        if !toModify.validate() {
+            throw TupletError.replacementNotSameDuration
+        } else {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: No need for `else`?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 4 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Yeah. I didn't want to have this be nil, since it should totally never happen. I feel like this one is ok to crash since it is so unlikely.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 6 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: I'm not sure what you mean exactly. This method can only replace a single note in a Tuplet with either a new Tuplet or a new note. I made it generic so that the caller can pass either a note or Tuplet here. I'm not sure what issue you are seeing. Can you try rephrasing maybe?

File: MusicNotationKit/MusicNotationKit/NoteCollection.swift (hunk 3 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/NoteCollection.swift]: Yeah. I will add a comment.

Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/NoteCollection.swift]: I could do that. I actually considered having this implement the Swift protocol that defines `first` and `last`, but that had way too much to it.

My reason for not wanting to make it optional is that it really isn't optional. It should never happen that `first` or `last` is optional, because you can not make a note less or more than a count of 1 and a Tuplet can only be instantiated with a valid set of notes, which must be greater than or equal to 1.

What do you think? I could make it optional just for safety, but for this, I feel like I could make the exception for throwing a `fatalError` since it's so impossible.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 9]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+                }
+            case nil:
+                break
+            }
+        }
+
+        if range.count == 1 {
+            let originalTie = try note(at: range.lowerBound).tie
+            try modifyState(forTie: originalTie)
+        } else {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: `return modifiedCollections` instead of `else`? Should we still try to return early in all cases?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 10]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+            let firstOriginalTie = try note(at: range.lowerBound).tie
+            let lastOriginalTie = try note(at: range.upperBound).tie
+
+            if firstOriginalTie == .beginAndEnd || lastOriginalTie == .beginAndEnd {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]:
```swift
guard firstOriginalTie != .beginAndEnd && lastOriginalTie != .beginAndEnd else {
  throw TupletError.invalidTieState
}
```

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 11]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+                throw TupletError.invalidTieState
+            } else if firstOriginalTie == .begin || lastOriginalTie == .end {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: from `else if` to 
```swift
if firstOriginalTie != .begin && lastOriginalTie != .end {
  try modifyState(forTie: firstOriginalTie)
  ...
}
```

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 8 above)
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]:
```swift
guard toModify.validate() else {
  throw TupletError.replacementNotSameDuration
}
self = toModify
```

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 12]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+        }
+        if !toModify.validate() {
+            throw TupletError.replacementNotSameDuration
+        } else {
+            self = toModify
+        }
+    }
+
+    private mutating func insert(_ noteCollections: [NoteCollection], at flatIndex: [Int]) throws {
+        if flatIndex.count != 1 {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: if `count == 1` return early after `notes.insert`?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 13]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+            // They are not compound tuplets
+            // Just remove from the notes
+            let removeRange = flatIndexes[0][0]...flatIndexes[flatIndexes.count - 1][0]
+            notes.removeSubrange(removeRange)
+        }
     }
 
-    internal mutating func replaceNotes(in range: Range<Int>, with noteCollection: NoteCollection) throws {
-        throw NSError()
+    private mutating func removeNotes(at flatIndexes: [[Int]]) throws {
```
Comment (migue48) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: If the `flatIndexes` are ordered, then we could traverse in reverse order  and create a `recursiveRemoveNote`  that removes single notes. If the result tuplet is empty, then we can remove the tuplet container. We can try to do this later if is too much trouble to do now.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift [hunk 14]
```diff
@@ -117,72 +132,338 @@ public struct Tuplet: NoteCollection {
...
+            return flatIndexes[index]
+        }
+
+        // If first - 1 (tuplet) != first (tuplet) && last + 1 (tuplet) != last (tuplet): return true;
+        // OR the count is 1 (not a nested tuplet)
+        // else return false
+        let beforeFirstFlatIndex = flatIndex(at: range.lowerBound - 1)
+        let afterLastFlatIndex = flatIndex(at: range.upperBound + 1)
+
+        switch (beforeFirstFlatIndex, afterLastFlatIndex) {
```
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Hmm. I tried what you are saying and it doesn't seem to cover everything correctly as the tests pass. Do you find this particularly difficult to reason through? Or do you think it's overly complicated? I think the switch statement and comment above makes it not too hard to follow.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 9 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: I changed it to a `guard` which I think basically does what you are saying. Not sure if it's more readable or not, but seems about the same.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 10 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Sounds good.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 8 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Yep. Sounds good.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 12 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Yep.

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 13 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: Yeah, I'm not sure I totally get you. Can you maybe make an issue for this as an enhancement with low priority?

File: MusicNotationKit/MusicNotationKit/Tuplet.swift (hunk 1 above)
Comment (drumnkyle) [inline MusicNotationKit/MusicNotationKit/Tuplet.swift]: The way it collapsed it, it doesn't show, but we are already comparing each note.
